faq.md

```diff
@@ -1,45 +1,634 @@
+
+### How do I create shaders?
+
+That’s outside the scope of this FAQ, but there’s a [tutorial on shaders and GLSL](https://opentk.net/learn/chapter1/4-shaders.html) as part of our lessons here, derived from the [tutorial on shaders and GLSL](https://learnopengl.com/Getting-started/Shaders) at [Learn OpenGL](https://learnopengl.com).
+
+### How do I handle errors?
+
+The same way as in OpenGL/CL/AL!  OpenTK exposes the same `glError()` functions, so you can test for errors in C# the same way you would in C.  Here’s a simple example:
+
+*Example code in C:*
+
+```c
+int shader = glCreateShader(GL_FRAGMENT_SHADER);
+
+int errorCode = glGetError();
+if (errorCode) {
+    printf("Uh oh.");
+    exit(-1);
+}
+```
+
+*C# equivalent:*
+
+```c#
+int shader = GL.CreateShader(ShaderType.FragmentShader);
+
+ErrorCode errorCode = GL.GetError();
+if (errorCode != ErrorCode.NoError)
+{
+    throw new UhOhException(...);
+}
+```
+
+Arguably, throwing exceptions is a better way to handle errors than aborting.
+
+`Gl.GetError()` in OpenTK reads errors from a queue, just like `glGetError()` does — it’s just a passthrough to the same underlying error state.
+
+#### That seems like a lot of if-statements; is there a better way?
+
+OpenGL 4.3+ offers a new technique, called a _debug message callback_.  (Many older OpenGL drivers also support this via the `KHR_debug` extension.)  The general idea behind this is to “register” a _debug callback_ function with OpenGL; any time OpenGL encounters an error, it will call your function immediately, and pass useful things to it like an error message.  First, you’ll need to declare an error-handling method that OpenGL can call, something like this:
+
+```c#
+private static void DebugCallback(DebugSource source, DebugType type, int id,
+    DebugSeverity severity, int length, IntPtr message, IntPtr userParam)
+{
+    string messageString = Marshal.PtrToStringAnsi(message, length);
+    Console.WriteLine($"{severity} {type} | {messageString}");
+
+    if (type == DebugType.DebugTypeError)
+        throw new Exception(messageString);
+}
+```
+
+Next, you’ll need to create a delegate from it as a handle that the GC can’t collect, and then register the resulting function with OpenGL:
+
+```c#
+private static DebugProc _debugProcCallback = DebugCallback;
+private static GCHandle _debugProcCallbackHandle;
+
+...
+    
+_debugProcCallbackHandle = GCHandle.Alloc(_debugProcCallback);
+
+GL.DebugMessageCallback(_debugProcCallback, IntPtr.Zero);
+GL.Enable(EnableCap.DebugOutput);
+GL.Enable(EnableCap.DebugOutputSynchronous);
+```
+
+Vassalware, a member of the OpenTK Team, has a [more extensive discussion of this technique](https://gist.github.com/Vassalware/d47ff5e60580caf2cbbf0f31aa20af5d).
```

Comment:
This section is taking up quite a big part of the FAQ and it's pretty in the weeds about implementation, I feel like it might be better to just make this it's own page and link to it here.
Then we can either take Vassals gist and maintain it ourselves or just link to it here.

Reply:
It's in the weeds, true, but error-handling is such an important topic for *any* library that I really think we should keep it.  OpenGL tutorials often gloss over error-handling; we should not, and should not just teach people how to use OpenGL, but how to use it *well*.  "How do I handle errors?" is a reasonable question that's part of "How do I use OpenGL well?" — and a good answer to it takes space.  We can afford a few extra bytes and lines of text to include a good answer.

Reply:
I agree that we should have an answer, but I think that answer can be a link to another page. Though idk if the website project would handle that gracefully.
We can leave it like this for this PR and see if we want to change it later.